contrib/epee/include/serialization/wire/write.h

```diff
@@ -0,0 +1,287 @@
+  {
+    try
+    {
+      W out{std::move(dest)};
+      bytes(out, source);
+      dest = out.take_buffer();
+    }
+    catch (const wire::exception& e)
+    {
+      dest.clear();
+      return e.code();
+    }
+    catch (...)
+    {
+      dest.clear();
+      throw;
+    }
+    return {};
+  }
+
+  template<typename T>
+  inline std::size_t array_size(std::true_type, const T& source)
+  { return boost::size(source); }
+
+  template<typename T>
+  inline constexpr std::size_t array_size(std::false_type, const T&) noexcept
+  { return 0; }
+
+  template<typename W, typename T>
+  inline void array(W& dest, const T& source)
+  {
+    using value_type = typename T::value_type;
+    static_assert(!std::is_same<value_type, char>::value, "write array of chars as string");
+    static_assert(!std::is_same<value_type, std::int8_t>::value, "write array of signed chars as binary");
+    static_assert(!std::is_same<value_type, std::uint8_t>::value, "write array of unsigned chars as binary");
+
+    dest.start_array(array_size(dest.need_array_size(), source));
+    for (const auto& elem : source)
+      bytes(dest, elem);
+    dest.end_array();
+  }
+
+  template<typename W, typename T>
+  inline bool field(W& dest, const wire::field_<T, true>& field)
+  {
+    // Arrays always optional, see `wire/field.h`
+    if (wire::available(field))
+    {
+      dest.key(field.name);
+      bytes(dest, field.get_value());
+    }
+    return true;
+  }
+
+  template<typename W, typename T>
+  inline bool field(W& dest, const wire::field_<T, false>& field)
+  {
+    if (wire::available(field))
+    {
+      dest.key(field.name);
+      bytes(dest, *field.get_value());
+    }
+    return true;
+  }
+
+  template<typename W, typename T>
+  inline std::enable_if_t<std::is_pod<T>::value> dynamic_object_key(W& dest, const T& source)
+  {
+    dest.binary_key(epee::as_byte_span(source));
+  }
+
+  template<typename W>
+  inline void dynamic_object_key(W& dest, const boost::string_ref source)
+  {
+    dest.key(source);
+  }
+
+  template<typename W, typename T>
+  inline void dynamic_object(W& dest, const T& source)
+  {
+    dest.start_object(source.size());
+    for (const auto& elem : source)
+    {
+      dynamic_object_key(dest, elem.first);
+      bytes(dest, elem.second);
+    }
+    dest.end_object();
+  }
+
+  template<typename W, typename... T>
+  inline void object(W& dest, T&&... fields)
+  {
+    dest.start_object(wire::sum(std::size_t(wire::available(fields))...));
+    const bool dummy[] = {field(dest, std::forward<T>(fields))...};
+    dest.end_object();
+    (void)dummy; // expand into array to get 0,1,2,etc order
+  }
+} // wire_write
+
+namespace wire
+{
+  template<typename W, typename T>
+  inline std::enable_if_t<is_array<T>::value> write_bytes(W& dest, const T& source)
+  {
+    wire_write::array(dest, source);
+  }
+
+  template<typename W, typename... T>
+  inline std::enable_if_t<std::is_base_of<writer, W>::value> object(W& dest, T... fields)
```

Comment:
Why is the `enable_if_t` needed here?

Reply:
There's an overload in `read.h` for the `wire::reader` case. I could omit this until the reader gets merged, but I think it's worth keeping in.

The overload allows for easy cpp usage - the `write_bytes` and `read_bytes` can forward to a single templated "map" function that calls `wire::object(...)` once for both the read and write case. i.e.

```c++
struct foo { int x; int y; }
void write_bytes(wire::writer&, const foo&);
void read_bytes(wire::reader&, foo&);

// .. cpp
template<typename F, typename T>
static void foo_map(F& format, T& self)
{ wire::object(format, WIRE_FIELD(x), WIRE_FIELD(y)); }

void write_bytes(wire::writer& dest, const foo& source) { foo_map(dest, source); }
void read_bytes(wire::reader& source, foo& dest) { foo_map(source, dest); } 
```

The boilerplate in the header and cpp can also further by macro'ed to make this a bit easier. This retains the "specify" once of the current method, while allowing for cpp files to reduce header bloat.

Reply:
And ! I forgot to mention why not `wire::object(reader& .. )` - templating allows forwarding of the derived type for performance reasons. Many of the calls within our codebase only actually use one type (p2p only uses epee, etc.), so specifying the derived makes more sense.